language/stdlib/modules/approved_payment.move

```diff
@@ -0,0 +1,97 @@
+// Module that allows a payee to approve payments with a cryptographic signature. The basic flow is:
+// (1) Payer sends `metadata` to the payee
+// (2) Payee signs `metadata` and sends 64 byte signature back to the payer
+// (3) Payer sends an approved payment to the payee by sending a transaction invoking `deposit`
+//     with payment metadata + signature. The transaction will abort if the signature check fails.
+// Note: approved payments are an accounting convenience/a courtesy mechansim for the payee, *not*
+// a hurdle that must be cleared for all payments to the payee. In addition, approved payments do
+// not have replay protection.
+address 0x0:
+module ApprovedPayment {
+    use 0x0::Libra;
+    use 0x0::LibraAccount;
+    use 0x0::Signature;
+    use 0x0::Transaction;
+    use 0x0::Vector;
+
+    // A resource to be published under the payee's account
+    resource struct T {
+        // 32 byte single Ed25519 public key whose counterpart must be used to sign the payment
+        // metadata. Note that this is different (and simpler) than the `authentication_key` used in
+        // LibraAccount::T, which is a hash of a public key + signature scheme identifier.
+        public_key: vector<u8>,
+        // TODO: events?
+    }
+
+    // Deposit `coin` in `payee`'s account if the `signature` on the payment metadata matches the
+    // public key stored in the `approved_payment` resource
+    public fun deposit<Token>(
+        approved_payment: &T,
+        payee: address,
+        coin: Libra::T<Token>,
+        metadata: vector<u8>,
+        signature: vector<u8>
+    ) {
+        // Sanity check of signature validity
+        Transaction::assert(Vector::length(&signature) == 64, 9001); // TODO: proper error code
+        // Cryptographic check of signature validity
+        Transaction::assert(
+            Signature::ed25519_verify(
+                signature,
+                *&approved_payment.public_key,
+                copy metadata
+            ),
+            9002, // TODO: proper error code
+        );
+        LibraAccount::deposit_with_metadata<Token>(payee, coin, metadata)
+    }
+
+    // Wrapper of `deposit` that withdraw's from the sender's balance and uses the top-level
+    // `ApprovedPayment.T` resource under the payee account.
```

Comment:
Nit: move syntax for the type.

Reply:
I keep doing this. Eventually I'll forget the IR...